projects/packages/waf/src/class-waf-request.php

```diff
@@ -103,4 +113,235 @@ private function get_ip_by_header( $headers ) {
 
 		return null;
 	}
+
+	/**
+	 * Returns the headers that were sent with this request
+	 *
+	 * @return array{ 0: string, 1: scalar }[]
+	 */
+	public function get_headers() {
+		$value              = array();
+		$has_content_type   = false;
+		$has_content_length = false;
+		foreach ( $_SERVER as $k => $v ) {
+			$k = strtolower( $k );
+			if ( 'http_' === substr( $k, 0, 5 ) ) {
+				$value[] = array( $this->normalize_header_name( substr( $k, 5 ) ), $v );
+			} elseif ( 'content_type' === $k && '' !== $v ) {
+				$has_content_type = true;
+				$value[]          = array( 'content-type', $v );
+			} elseif ( 'content_length' === $k && '' !== $v ) {
+				$has_content_length = true;
+				$value[]            = array( 'content-length', $v );
+			}
+		}
+		if ( ! $has_content_type ) {
+			// default Content-Type per RFC 7231 section 3.1.5.5.
+			$value[] = array( 'content-type', 'application/octet-stream' );
+		}
+		if ( ! $has_content_length ) {
+			$value[] = array( 'content-length', '0' );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Change a header name to all-lowercase and replace spaces and underscores with dashes.
+	 *
+	 * @param string $name The header name to normalize.
+	 * @return string
+	 */
+	public function normalize_header_name( $name ) {
+		return str_replace( array( ' ', '_' ), '-', strtolower( $name ) );
+	}
+
+	/**
+	 * Get the method for this request (GET, POST, etc).
+	 *
+	 * @return string
+	 */
+	public function get_method() {
+		return isset( $_SERVER['REQUEST_METHOD'] )
+			? filter_var( wp_unslash( $_SERVER['REQUEST_METHOD'] ), FILTER_DEFAULT )
+			: '';
+	}
+
+	/**
+	 * Get the protocol for this request (HTTP, HTTPS, etc)
+	 *
+	 * @return string
+	 */
+	public function get_protocol() {
+		return isset( $_SERVER['SERVER_PROTOCOL'] )
+			? filter_var( wp_unslash( $_SERVER['SERVER_PROTOCOL'] ), FILTER_DEFAULT )
+			: '';
+	}
```

Comment:
Previously this was returning whether or not HTTPS was used in the request (i.e. `$_SERVER['HTTPS']`), but in ModSecurity `REQUEST_PROTOCOL` is the HTTP protocol that was used with the request (see CRS rules `920180` and `920430`).